Add pattern matching documentation

zverok opened the pull request:
I am not sure that what I am doing is necessary, but I noticed Ruby 2.7 release lacks any pattern matching documentation.

Maybe it is intentional (to underline features experimentality?), but it probably can harm the adoption.
So, trying to document the feature to the best of my understanding/experimenting.

zverok commented on doc/syntax/control_expressions.rdoc:
```diff
@@ -232,7 +232,7 @@ You may use +then+ after the +when+ condition.  This is most frequently used
 
   case a
-  when 1, 2 then puts "a is one or two
+  when 1, 2 then puts "a is one or two"
```
It was a typo here before...

ko1 commented on doc/syntax/control_expressions.rdoc:
```diff
@@ -255,6 +255,18 @@ Again, the +then+ and +else+ are optional.
 The result value of a +case+ expression is the last value executed in the
 expression.
 
+Since Ruby 2.7, +case+ expressions also provide more powerful pattern matching feature with +in+ keyword:
```
add "experimental feature"

ko1 commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,364 @@
+= Pattern matching
+
+Pattern matching is a feature allowing deep matching of structured values, with checking the structure and binding matched parts to local variables.
```
add "experimental feature"

ko1 commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,364 @@
+
+== Current feature status
+
+As of Ruby 2.7, feature is considered _experimental_: its syntax can (slightly) change in the future, and the performance is not optimized yet. Every time you use pattern matching in code, the warning will be printed:
```
slightly?

zverok replied:
I was trying to communicate that it is unlikely for the syntax to change completely in upcoming versions, but it is probably inappropriate here. Removed.

zverok commented:
@ko1 Thanks for the review. I've addressed your comments (and also expanded explanations a bit, providing introductory examples of pattern matching use).

palkan commented on doc/syntax.rdoc:
```diff
@@ -11,6 +11,9 @@ Assignment[rdoc-ref:syntax/assignment.rdoc] ::
   +if+, +unless+, +while+, +until+, +for+, +break+, +next+, +redo+
 
+{Pattern matching}[rdoc-ref:syntax/pattern_matching.rdoc] ::
+  Experimental structure pattern matching and variable binding syntax
```
"structural" (e.g., [here](http://blog.higher-order.com/blog/2009/08/21/structural-pattern-matching-in-java/) and [here](https://blog.carbonfive.com/2018/01/08/an-introduction-to-adts-and-structural-pattern-matching-in-typescript/)) ? And probably the word "feature" is missing here, too.

```suggestion
  Experimental structural pattern matching and variables binding feature syntax
```

zverok replied:
Agreed about "structural" (fixed), but I don't believe "feature" is suitable here (in generic "TOC of syntax elements").

palkan commented on doc/syntax/control_expressions.rdoc:
```diff
@@ -255,6 +255,19 @@ Again, the +then+ and +else+ are optional.
 The result value of a +case+ expression is the last value executed in the
 expression.
 
+Since Ruby 2.7, +case+ expressions also provide more powerful experimental
```
```suggestion
Since Ruby 2.7, +case+ expressions also provide a more powerful experimental
```

palkan commented on doc/syntax/control_expressions.rdoc:
```diff
@@ -255,6 +255,19 @@ Again, the +then+ and +else+ are optional.
 expression.
 
+Since Ruby 2.7, +case+ expressions also provide more powerful experimental
+pattern matching feature with +in+ keyword:
```
```suggestion
pattern matching feature via the +in+ keyword:
```

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+= Pattern matching
+
+Pattern matching is an experimental feature allowing deep matching of structured values, with checking the structure and binding matched parts to local variables.
```
How about?
```suggestion
Pattern matching is an experimental feature allowing deep matching of structured values: checking the structure, and binding the matched parts to local variables.
```

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+
+Pattern matching is an experimental feature allowing deep matching of structured values, with checking the structure and binding matched parts to local variables.
+
+Pattern matching in Ruby is implemented with +in+ operator, which can be used in a separate expression:
```
```suggestion
Pattern matching in Ruby is implemented with the +in+ operator, which can be used in a standalone expression:
```

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+
+    <variable> in <pattern>
+
+or in +case+ statement:
```
```suggestion
or within the +case+ statement:
```

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+Array patterns match arrays, or objects that respond to +deconstruct+ (see below about the latter).
+Hash patterns match hashes, or objects that respond to +deconstruct_keys+ (see below about the latter). Note that only symbol keys are supported for hash patterns, at least for now.
+
+Important difference between array and hash pattern behavior is arrays match only a _whole_ array
```
Maybe, we need to explain the "whole" word a bit, e.g.:

```suggestion
Important difference between array and hash patterns behaviour is that the former match arrays as a _whole_ (i.e., every element _must_ match):
```

zverok replied:
I don't think "every element must match" adds any clarity (for me, it makes me thin, "oh, maybe every element must match the same pattern, or what?.."), I believe the provided example explains the meaning just fine.

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+  end
+  #=> "not matched"
+
+while the hash matches even if there are other keys besides specified part:
```
Shouldn't we mention the `**nil` pattern which makes the hash pattern _strict_?

zverok replied:
Ugh, definitely! Thanks, added.

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+  end
+  #=> "matched"
+
+In +case+ (but not in standalone +in+ statement), parentheses around both kinds of patterns could be omitted:
```
```suggestion
In +case+ (but not in standalone +in+) statement, parentheses around both kinds of patterns could be omitted
```

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+
+== Variable binding
+
+Besides deep structural checks, one of very important features of the pattern matching is binding matched parts to a local variables. The basic form of binding is just specifying <code>=> variable_name</code> after the matched (sub)pattern (one might find this similar to storing exceptions in local variables in <code>rescue ExceptionClass => var</code> clause):
```
```suggestion
Besides deep structural checks, one of the very important features of the pattern matching is the binding of the matched parts to local variables. The basic form of binding is just specifying <code>=> variable_name</code> after the matched (sub)pattern (one might find this similar to storing exceptions in local variables in <code>rescue ExceptionClass => var</code> clause):
```

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+  end
+  #=> "matched: 1"
+
+If there is no additional check required, just binding some part of the data to a variable, simpler form may be used:
```
```suggestion
If no additional check is required, only binding some part of the data to a variable, a simpler form could be used:
```

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+  end
+  #=> "matched: 1"
+
+For hash patterns, even simpler form exists: just key specification (without any value) binds the local variable with the key's name, too:
```
```suggestion
For hash patterns, even a simpler form exists: key-only specification (without any value) binds the local variable with the key's name, too:
```

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+  end
+  #=> "matched: 1"
+
+Binding works for nested patterns:
```
```suggestion
Binding works for nested patterns as well:
```

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+  end
+  #=> "matched: Jane"
+
+The "rest" part of pattern also can be bound to a variable:
```
```suggestion
The "rest" part of a pattern also can be bound to a variable:
```

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+  end
+  #=> "matched. school: 2"
+
+  case john # he specified his school level is "high", but last school does not match
```
```suggestion
  case john # the specified school level is "high", but last school does not match
```

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+
+== Matching non-primitive objects: +deconstruct_keys+ and +deconstruct+
+
+As already mentioned above, hash and array patterns besides literal arrays and hashes will try to match any object, if it is implementing +deconstruct+ (for array pattern) or +deconstruct_keys+ (for hash pattern).
```
```suggestion
As already mentioned above, hash and array patterns besides literal arrays and hashes will try to match any object implementing +deconstruct+ (for array patterns) or +deconstruct_keys+ (for hash patterns).
```

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+  # prints: deconstruct_keys called with [:x]
+  #=> "matched: 1"
+
++keys+ are passed to +deconstruct_keys+ to provide room for optimization in matched class: if calculating full hash representation is expensive, it may calculate only the necessary subhash.
```
```suggestion
+keys+ are passed to +deconstruct_keys+ to provide a room for optimization in the matched class: if calculating a full hash representation is expensive, one may calculate only the necessary subhash.
```

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+
++keys+ are passed to +deconstruct_keys+ to provide room for optimization in matched class: if calculating full hash representation is expensive, it may calculate only the necessary subhash.
+
+Additionally, when matching custom classes, expected class could be specified as part of the pattern and is checked with <code>===</code>
```
```suggestion
Additionally, when matching custom classes, expected class could be specified as a part of the pattern and is checked with <code>===</code>
```

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+
+== Guard clauses
+
++if+ can be used to attach additional condition (guard clause) when matched pattern
```
```suggestion
+if+ can be used to attach an additional condition (guard clause) when the pattern matches
```

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+
+Alternatively, command-line key <code>-W:no-experimental</code> can be used to turn off "experimental" feature warnings.
+
+One of the things developer should be aware about, which probably to be fixed in upcoming versions, is that pattern matching statement rewrites mentioned local variables on partial match, <i>even if the whole pattern is not matched</i>.
```
```suggestion
One of the things developer should be aware of, which probably to be fixed in the upcoming versions, is that pattern matching statement rewrites mentioned local variables on partial match, <i>even if the whole pattern is not matched</i>.
```

palkan commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,398 @@
+  a
+  #=> 1  -- the whole pattern not matched, but partial match happened, a is rewritten
+
+Of the core classes, the only currently implementing +deconstruct+ and +deconstruct_keys+ is Struct.
```
```suggestion
Currently, the only core class implementing +deconstruct+ and +deconstruct_keys+ is Struct.
```

zverok commented:
@palkan Thanks a lot! Applied most of your suggestions, answered the rest :)

zverok commented:
@k0kubun @k-tsj Any news on this one?..
It would be probably good to have pattern-matching docs _somewhere_?..

k0kubun commented:
Please wait for the mainainer of pattern matching, not me.

k-tsj commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,412 @@
+  # prints: deconstruct_keys called with [:x]
+  #=> "matched: 1"
+
++keys+ are passed to +deconstruct_keys+ to provide a room for optimization in the matched class: if calculating a full hash representation is expensive, one may calculate only the necessary subhash.
```
Note: `keys` will be `nil` when a rest variable is specified in the pattern.

```ruby
class C
  def deconstruct_keys(keys)
    p keys
   {}
  end
end
C.new in {**rest}
```

k-tsj commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,412 @@
+
+To suppress this warning, one may use newly introduced Warning::[]= method:
+
+  Warning[:experimental] = false
```
"experimental" warnings for pattern matching being emitted in compile time, so this example code doesn't suppress the warning.

`Warning::[]=` is useful for following case.

````
$ cat a.rb
Warning[:experimental] = false
require_relative 'b.rb'

$ cat b.rb
{a: 1, b: 2} in {a:}
```

zverok commented:
@k-tsj thank you for your review I addressed all the comments.

> I think this document is good introduction for beginners.
> In addition to this, it would be nice to provide reference manual like other languages.

I believe (at least this was my intention) that the document is not an introduction but fully covers the pattern matching feature. If there are some aspects that are not covered currently, please kindly point at them.

I agree, though, that it explains the feature in a "walkthrough"/showcase way, not very formal but allowing to quickly grasp all the details. I believe most of current Ruby's docs are written this way (but to the best of my knowledge, no style guide exists currently).

Whether this documentation should be made more formal, with grammar specification and more sections, I am really not sure, and probably not ready to invest my time in.

baweaver commented on doc/syntax/control_expressions.rdoc:
```diff
@@ -255,6 +255,19 @@ Again, the +then+ and +else+ are optional.
+  end
+  # => "matched: 1"
+
+Pattern matching syntax is described on {its own page}[rdoc-ref:syntax/pattern_matching.rdoc].
```
Consider adding "The" before this.

baweaver commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,441 @@
+
+* any Ruby object (matched by <code>===</code> operator, like in +when+);
+* array pattern: <code>[<subpattern>, <subpattern>, <subpattern>, ...]</code>;
+* hash pattern: <code>{key: <subpattern>, key: <supattern>, ...}</code>;
```
Typo: supattern

baweaver commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,441 @@
+Array patterns match arrays, or objects that respond to +deconstruct+ (see below about the latter).
+Hash patterns match hashes, or objects that respond to +deconstruct_keys+ (see below about the latter). Note that only symbol keys are supported for hash patterns, at least for now.
+
+Important difference between array and hash patterns behavior is arrays match only a _whole_ array
```
`An important difference...`

baweaver commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,441 @@
+  end
+  #=> "matched a whole"
+
+Both array and hash patterns support "rest" specification:
```
It may be beneficial to mention that "rest" patterns can be captured as well:

```ruby
case [1, 2, 3]
in [Integer, *rest] then rest
else false
end
=> [2, 3]

case { a: 1, b: 2, c: 3 }
in { a: Integer, **rest } then rest
else false
end
=> {:b=>2, :c=>3}
```

zverok replied:
All the capturing (including `**rest`) [explained](https://github.com/ruby/ruby/pull/2786/files/4698f97f0e4bcd23f0a57cab7ffab7e8b7e16e0d#diff-63b14a593bb03694f704ae5bc20845c7R203) in the next section -- "Variable binding".

baweaver commented:
How much do we want copy editing on this for english-isms?

baweaver commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,441 @@
+  end
+
+  case Point.new(1, -2)
+  in SuperPoint(x: 0.. => px)
```
This will raise an exception as there's  no `SuperPoint` method

palkan replied:
This is a valid syntax for the _constant pattern_ (see [ruby/spec](https://github.com/ruby/ruby/blob/2dea81f47728b9a28831dd91fc14dafcc3d29271/spec/ruby/language/pattern_matching_spec.rb#L673)).

zverok replied:
This (both `()` and `[]`) is just a part of the pattern-matching syntax, it has nothing to do with method definitions. This just works:
```ruby
class Point
  def self.[](*)
    raise ArgumentError, 'Fuuu'
  end

  def initialize(x, y)
    @x, @y = x, y
  end

  def deconstruct
    puts "deconstruct called"
    [@x, @y]
  end

  def deconstruct_keys(keys)
    puts "deconstruct_keys called with #{keys.inspect}"
    {x: @x, y: @y}
  end
end

Point.new(1, 2) in Point(Integer, Integer) # no method Point() is necessary
Point.new(1, 2) in Point[Integer, Integer] # method Point.[] is not called
```

baweaver commented on doc/syntax/pattern_matching.rdoc:
```diff
@@ -0,0 +1,441 @@
+
+== Guard clauses
+
++if+ can be used to attach an additional condition (guard clause) when the pattern matches
```
Missing a `:` on the end. You might also mention that it has access to the variables that are captured

zverok commented:
@baweaver Thanks for your review, I made amendments.
(And don't hesitate to fix my "Ukrainglish", I know it is peculiar and is better to be fixed by the native speaker, mercilessly!)

k-tsj commented:
OK.

The current document covers pattern matching feature enough, so I'd like to merge this PR to get started.
(In the future, we may split this document into a brief guide by walkthrough/showcase way, and a formal reference.)

Could you squash commits into one?

zverok commented:
@k-tsj Thank you. Commits are squashed & the branch rebased from the master.

k-tsj commented:
Merged.

Thank you for your hard work.

zverok commented:
@k-tsj Thank you!
Do you know what is necessary for new docs to be rendered at https://docs.ruby-lang.org/en/master/ ?
I tended to believe it always renders the docs from current `master`, but I am not sure now.

CC @znz @hsbt
